internal/controller/reinstall/reinstall.go

```diff
@@ -372,3 +379,279 @@ func (r *ReinstallHandler) finalizeReinstallRequest(
+			}
+		}
+	}()
+
+	if clusterInstance.Spec.Reinstall.PreservationMode == v1alpha1.PreservationModeNone {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		return
+	}
+
+	preservationDataBackedupCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallPreservationDataBackedup)
+	clusterIdentityDetectedCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallClusterIdentityDataDetected)
+
+	setInProgress := false
+	if preservationDataBackedupCondition == nil ||
+		preservationDataBackedupCondition.Reason == string(v1alpha1.Initialized) {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if clusterIdentityDetectedCondition == nil ||
+		clusterIdentityDetectedCondition.Reason == string(v1alpha1.Initialized) {
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if setInProgress {
+		result.Requeue = true
+		return
+	}
+
+	// Determine if resources have been preserved already
+	preservationExecuted := false
+	switch preservationDataBackedupCondition.Reason {
+	case string(v1alpha1.Failed):
+		err = errors.New(preservationDataBackedupCondition.Message)
+		return
+
+	case string(v1alpha1.Completed), string(v1alpha1.DataUnavailable):
+		preservationExecuted = true
+	}
+
+	clusterIDDataDetectionExecuted := false
+	switch clusterIdentityDetectedCondition.Reason {
+	case string(v1alpha1.Failed):
+		err = errors.New(clusterIdentityDetectedCondition.Message)
+		return
+
+	case string(v1alpha1.DataAvailable), string(v1alpha1.DataUnavailable):
+		clusterIDDataDetectionExecuted = true
+	}
+
+	if preservationExecuted && clusterIDDataDetectionExecuted {
+		return
+	}
+
+	log.Info("Starting data preservation")
+
+	err = preservation.Backup(ctx, r.Client, log, clusterInstance)
+	if err != nil {
+		log.Error("Encountered error during data preservation", zap.Error(err))
+
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.Failed, err.Error())
+
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.Failed, "Data preservation failed")
+
+		return
+	}
+
+	log.Info("Data preservation completed, computing backup summary...")
+
+	// Add a small delay to allow for the last resource preserved to be created before computing stats
+	time.Sleep(1 * time.Second)
```

Comment:
Did you actually run into an issue that required you to do this?

If you got a successful status from the API saying an object was created you should be safe to continue.

Reply:
I was pre-anticipating potential network / server delays in object creation. I've removed these for now, should an issue arise in the future, we can revisit this.